### PR TITLE
internal/charmstore: store SHA256 hash too

### PR DESCRIPTION
**Pull request by rogpeppe**
We need SHA256 for the legacy API.

### CODE REVIEW

**Comment by jujugui** [comment]
Test FAILed.
Refer to this link for build results: http://ci.jujugui.org:8080/job/charmstore/216/

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results: http://ci.jujugui.org:8080/job/charmstore/217/

**Comment by jrwren** [comment]
:+1:

File: internal/v4/archive.go [hunk 1]
```diff
@@ -114,10 +116,13 @@ func (h *Handler) servePostArchive(id *charm.Reference, w http.ResponseWriter, r
 	// Upload the actual blob, and make sure that it is removed
 	// if we fail later.
 	name := bson.NewObjectId().Hex()
-	err = h.store.BlobStore.PutUnchallenged(req.Body, name, req.ContentLength, hash)
+	hash256 := sha256.New()
+	reader := io.TeeReader(req.Body, hash256)
```
Comment (frankban) [inline internal/v4/archive.go:120]: Nice!

**Comment by frankban** [comment]
:+1:

**Comment by rogpeppe** [comment]
superceded by #100

**Pull request closed without merging**